lib/chatter/message_filter.ex

```diff
@@ -0,0 +1,106 @@
+defmodule Chatter.MessageFilter do
+  @moduledoc """
+  Describe messages filtering functionality
+  """
+
+  alias Chatter.{MessageAgent, Search}
+
+  def filter_by_params({search, nil = _filter_option}), do: get_search_messages(search)
+
+  def filter_by_params({search, filter_option}) do
+    search_messages = get_search_messages(search)
+    filter_option
+      |> filter_by_option()
+      |> Enum.filter(fn message -> message in search_messages end)
+  end
+
+  def filter_by_option(filter_option) do
+    case filter_option do
+      :with_likes_who_liked -> filter_with_likes_who_like()
+      :without_likes_who_never_liked -> filter_without_likes_who_never_liked()
+      :with_major_likes -> filter_with_major_likes()
+    end
+  end
+
+  def get_search_messages(search), do: Enum.filter(MessageAgent.get(), &filter(&1, search))
+
+  def filter(message, search) do
+    filter_text(message, search) && filter_likes(message, search)
+  end
+
+  defp filter_text(message, %Search{text: nil} = _search), do: message
+
+  defp filter_text(message, %Search{text: search_text} = _search), do: message.text =~ search_text
+
+  defp filter_likes(message, %{likes: nil} = _search), do: message
+
+  defp filter_likes(message, %{likes_option: option, likes: likes_count} = _search) do
+    case option do
+      ">=" -> Enum.count(message.likes) >= likes_count
+      "<=" -> Enum.count(message.likes) <= likes_count
+      "=" -> Enum.count(message.likes) == likes_count
+    end
+  end
+
+  defp filter_with_likes_who_like do
+    all_likes = MessageAgent.get_all_likes()
```

Comment:
hardcoding agent usage is a bad thing
it is better to pass it as arguments

Reply:
right, messages data moved to params